core/src/test/scala/org/apache/spark/sql/delta/skipping/MultiDimClusteringSuite.scala

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.skipping
+
+import java.io.File
+
+import scala.util.Random
+
+// scalastyle:off import.ordering.noEmptyLine
+import org.apache.spark.sql.delta.sources.DeltaSQLConf._
+import org.apache.spark.sql.delta.test.DeltaSQLCommandTest
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.functions.expr
+import org.apache.spark.sql.test.SharedSparkSession
+
+class MultiDimClusteringSuite extends QueryTest
+    with SharedSparkSession with DeltaSQLCommandTest {
+
+  private lazy val sparkSession = spark
+  import sparkSession.implicits._
+
+  test("Negative case - ZOrder clustering expression with zero columns") {
+    val ex = intercept[AssertionError] {
+      ZOrderClustering.getClusteringExpression(Seq.empty, 20)
+    }
+    assert(ex.getMessage contains "Cannot do Z-Order clustering by zero columns!")
+  }
+
+  test("ZOrder clustering expression with one column") {
+    val cluster = ZOrderClustering.getClusteringExpression(Seq(expr("col1")), 20)
+    assert(cluster.expr.toString ===
+      "cast(interleavebits(rangepartitionid('col1, 20)) as string)")
+  }
+
+  test("ZOrder clustering expression with two column") {
+    val cluster = ZOrderClustering.getClusteringExpression(Seq(expr("col1"), expr("col2")), 20)
+    assert(cluster.expr.toString ===
+      "cast(interleavebits(rangepartitionid('col1, 20), rangepartitionid('col2, 20)) as string)")
+  }
+
+  test("ensure records with close Z-order values are close in the output") {
+    withTempDir { tempDir =>
+      withSQLConf(
+        MDC_NUM_RANGE_IDS.key -> "4",
+        MDC_ADD_NOISE.key -> "false") {
+        val data = Seq(
+          "a" -> 20, "a" -> 20, // Same Z-order value // Z-Order 0
+          "b" -> 20,                                  // Z-Order 1
+          "c" -> 30,                                  // Z-Order 2
+          "d" -> 70,                                  // Z-Order 3
+          "e" -> 90, "e" -> 90, "e" -> 90, // Same Z-order value // Z-Order 5
```

Comment:
what does "Z-order X" mean? for example, why isn't there a Z-Order 4?

Reply:
It is just a number to indicate the order of record based on the Z-order value. Helps in understanding the expected results in the end of the test.